synapse/handlers/appservice.py

```diff
@@ -147,6 +169,60 @@ def query_room_alias_exists(self, room_alias):
                 defer.returnValue(result)
 
     @defer.inlineCallbacks
+    def query_3pu(self, protocol, fields):
+        services = yield self._get_services_for_3pn(protocol)
+
+        deferreds = []
+        for service in services:
+            deferreds.append(self.appservice_api.query_3pu(
+                service, protocol, fields
+            ))
+
+        results = yield defer.DeferredList(deferreds, consumeErrors=True)
+
+        ret = []
+        for (success, result) in results:
+            if not success:
+                continue
+            if not isinstance(result, list):
+                continue
```

Comment:
More logging added, and also moved into the ApplicationServiceApi class, where arguably it belongs (as IRL'ed). Done 6520163